Userland/Libraries/LibDisassembly/riscv64/Encoding.cpp

```diff
@@ -0,0 +1,549 @@
+/*
+ * Copyright (c) 2023, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Encoding.h"
+#include <AK/Assertions.h>
+#include <AK/BitCast.h>
+
+namespace Disassembly::RISCV64 {
+
+// rd is always in bit positions [11:7].
+static Register extract_destination_register(u32 instruction)
+{
+    auto register_bits = (instruction >> 7) & 0b11111;
+    return static_cast<Register>(register_bits);
+}
+
+// rs1 is always in bit positions [19:15].
+static Register extract_source_register_1(u32 instruction)
+{
+    auto register_bits = (instruction >> 15) & 0b11111;
+    return static_cast<Register>(register_bits);
+}
+
+// rs2 is always in bit positions [24:20].
+static Register extract_source_register_2(u32 instruction)
+{
+    auto register_bits = (instruction >> 20) & 0b11111;
+    return static_cast<Register>(register_bits);
+}
+
+static Register extract_compressed_rs2(u16 instruction)
+{
+    auto register_bits = (instruction >> 2) & 0b11111;
+    return static_cast<Register>(register_bits);
+}
+
+// The following register formats from C only use 3 bits and offset the register number by 8:
+
+// Used as rd/rs1 in CA.
+[[maybe_unused]] static Register extract_compressed_short_rs1(u16 instruction)
+{
+    auto register_bits = (instruction >> 7) & 0b111;
+    return static_cast<Register>(register_bits + 8);
+}
+
+// Used as rs2 in CS and CA, but as rd in CL and CIW.
+[[maybe_unused]] static Register extract_compressed_short_rs2(u16 instruction)
+{
+    auto register_bits = (instruction >> 2) & 0b111;
+    return static_cast<Register>(register_bits + 8);
+}
+
+static u8 extract_funct3(u32 instruction)
+{
+    return static_cast<u8>((instruction >> 12) & 0b111);
+}
+
+static u8 extract_funct7(u32 instruction)
+{
+    return static_cast<u8>((instruction >> 25) & 0b1111111);
+}
+
+// Only used for CB format to detect C.ANDI.
+[[maybe_unused]] static u8 extract_compressed_funct2_cb(u16 instruction)
+{
+    return static_cast<u8>((instruction >> 10) & 0b11);
+}
+
+[[maybe_unused]] static u8 extract_compressed_funct2(u16 instruction)
+{
+    return static_cast<u8>((instruction >> 5) & 0b11);
+}
+
+[[maybe_unused]] static u8 extract_compressed_funct4(u16 instruction)
+{
+    return static_cast<u8>((instruction >> 12) & 0b1111);
+}
+
+[[maybe_unused]] static u8 extract_compressed_funct6(u16 instruction)
+{
+    return static_cast<u8>((instruction >> 10) & 0b111111);
+}
+
+constexpr i32 sign_extend(u32 n, u8 size)
+{
+    // negative
+    if ((n & (1 << (size - 1))) > 0) {
+        return static_cast<i32>(n | (0xffffffff << size));
+    }
+    // positive
+    return static_cast<i32>(n);
```

Comment:
While probably well defined, I like `bit_cast` more in this case,
also you can omit the braces and `>0`

Reply:
I like doing these comparisons explicitly without relying on the int-> bool conversion